Disable network access for user `default` in docker images

Felixoid opened the pull request:
<!---
A technical comment, you are free to remove or leave it as it is when PR is created
The following categories are used in the next scripts, update them accordingly
utils/changelog/changelog.py
tests/ci/cancel_and_rerun_workflow_lambda/app.py
-->
### Changelog category (leave one):
- Build/Testing/Packaging Improvement


### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
*Potentially breaking*: Improvement to set even more restrictive defaults. The current defaults are already secure - the user has to specify an option to publish ports explicitly. But when the `default` user doesn’t have a password set by `CLICKHOUSE_PASSWORD` and/or a username changed by `CLICKHOUSE_USER` environment variables, it should be available only from the local system as an additional level of protection.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

<!---
Directly edit documentation source files in the "docs" folder with the same pull-request as code changes

or

Add a user-readable short description of the changes that should be added to docs.clickhouse.com below.

At a minimum, the following information should be added (but add more as needed).
- Motivation: Why is this function, table engine, etc. useful to ClickHouse users?

- Parameters: If the feature being added takes arguments, options or is influenced by settings, please list them below with a brief explanation.

- Example use: A query or command.
-->


> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing)

All builds in Builds_1 and Builds_2 stages are always mandatory and will run independently of the checks below:
- [ ] <!---ci_include_stateless--> Only: Stateless tests
- [ ] <!---ci_include_integration--> Only: Integration tests
- [ ] <!---ci_include_performance--> Only: Performance tests
---
- [ ] <!---ci_exclude_style--> Skip: Style check
- [ ] <!---ci_exclude_fast--> Skip: Fast test
---
- [ ] <!---woolen_wolfdog--> Run all checks ignoring all possible failures (Resource-intensive. All test jobs execute in parallel).
- [ ] <!---no_ci_cache--> Disable CI cache

Felixoid committed:
Disable network access for user `default` in docker images

robot-ch-test-poll4 commented:
<!-- automatic status comment for PR #75259 from ClickHouse/ClickHouse:docker-localhost-default -->
*This is an automated comment for commit 65435a3db7214261b793acfb69388567b4c8c9b3 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/75259/65435a3db7214261b793acfb69388567b4c8c9b3/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Builds</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/3bdc7ecfbc9f5562748c12f005be189780e643d9/builds/report.html">✅ success</a></td></tr>
<tr><td>ClickBench</td><td>Runs <a href="https://github.com/ClickHouse/ClickBench/">ClickBench</a> with instant-attach table</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/c08896ddcc9178b693f88064c16c905660b325cb/clickbench__aarch64_.html">✅ success</a></td></tr>
<tr><td>Compatibility check</td><td>Checks that <b>clickhouse</b> binary runs on distributions with old libc versions. If it fails, ask a maintainer for help</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/c08896ddcc9178b693f88064c16c905660b325cb/compatibility_check__aarch64_.html">✅ success</a></td></tr>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/4ff4869e3bdd97f7de69f51f994fc7089c6fd412/docs_check.html">✅ success</a></td></tr>
<tr><td>Fast test</td><td>Normally this is the first check that is ran for a PR. It builds ClickHouse and runs most of <a href="https://clickhouse.com/docs/en/development/tests#functional-tests">stateless functional tests</a>, omitting some. If it fails, further checks are not started until it is fixed. Look at the report to see which tests fail, then reproduce the failure locally as described <a href="https://clickhouse.com/docs/en/development/tests#functional-test-locally">here</a></td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/3bdc7ecfbc9f5562748c12f005be189780e643d9/fast_test.html">✅ success</a></td></tr>
<tr><td>Install packages</td><td>Checks that the built packages are installable in a clear environment</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/c08896ddcc9178b693f88064c16c905660b325cb/install_packages__aarch64_.html">✅ success</a></td></tr>
<tr><td>Integration tests</td><td>The integration tests report. In parenthesis the package type is given, and in square brackets are the optional part/total tests</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/c08896ddcc9178b693f88064c16c905660b325cb/integration_tests__aarch64__[1_6].html">✅ success</a></td></tr>
<tr><td>Performance Comparison</td><td>Measure changes in query performance. The performance test report is described in detail <a href="https://github.com/ClickHouse/ClickHouse/tree/master/docker/test/performance-comparison#how-to-read-the-report">here</a>. In square brackets are the optional part/total tests</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/c08896ddcc9178b693f88064c16c905660b325cb/performance_comparison__release__[1_4]/report.html">✅ success</a></td></tr>
<tr><td>Unit tests</td><td>Runs the unit tests for different release types</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/c08896ddcc9178b693f88064c16c905660b325cb/unit_tests__asan_.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

maxknv commented:
Why are these network restrictions needed for the "default" user but not needed for other users and vice versa?

darkleaf commented:
I've got failed CI and now I have to fix it
![meme-template](https://github.com/user-attachments/assets/30a5d503-4a2a-4634-ae31-36375d7d136e)

slvrtrn commented:
This change makes the DX significantly worse. Instead of simply running CH in Docker and then using `curl`/`clickhouse-client` to experiment with it, now the user needs to add `CLICKHOUSE_USER` + `CLICKHOUSE_PASSWORD` there.
It also breaks CI for language clients and other integrations, and it took some time to find the root cause of the error message, which is incredibly cryptic.

Felixoid commented:
Dear colleagues,
As much as I agree that this change decreases usability a little bit, our image has much less strict rules to launch than other DBs.

For example, when you try to launch `mariadb`, `mysql`, or `postgres` official images, they simply refuse to launch without preset ENVs:

```
$ docker run mariadb
2025-02-04 16:40:54+00:00 [Note] [Entrypoint]: Entrypoint script for MariaDB Server 1:11.6.2+maria~ubu2404 started.
2025-02-04 16:40:54+00:00 [Warn] [Entrypoint]: /sys/fs/cgroup///memory.pressure not writable, functionality unavailable to MariaDB
2025-02-04 16:40:54+00:00 [Note] [Entrypoint]: Switching to dedicated user 'mysql'
2025-02-04 16:40:54+00:00 [Note] [Entrypoint]: Entrypoint script for MariaDB Server 1:11.6.2+maria~ubu2404 started.
2025-02-04 16:40:55+00:00 [ERROR] [Entrypoint]: Database is uninitialized and password option is not specified
        You need to specify one of MARIADB_ROOT_PASSWORD, MARIADB_ROOT_PASSWORD_HASH, MARIADB_ALLOW_EMPTY_ROOT_PASSWORD and MARIADB_RANDOM_ROOT_PASSWORD
$ docker run mysql
2025-02-04 16:41:04+00:00 [Note] [Entrypoint]: Entrypoint script for MySQL Server 9.2.0-1.el9 started.
2025-02-04 16:41:05+00:00 [Note] [Entrypoint]: Switching to dedicated user 'mysql'
2025-02-04 16:41:05+00:00 [Note] [Entrypoint]: Entrypoint script for MySQL Server 9.2.0-1.el9 started.
2025-02-04 16:41:05+00:00 [ERROR] [Entrypoint]: Database is uninitialized and password option is not specified
    You need to specify one of the following as an environment variable:
    - MYSQL_ROOT_PASSWORD
    - MYSQL_ALLOW_EMPTY_PASSWORD
    - MYSQL_RANDOM_ROOT_PASSWORD
$ docker run postgres
Error: Database is uninitialized and superuser password is not specified.
       You must specify POSTGRES_PASSWORD to a non-empty value for the
       superuser. For example, "-e POSTGRES_PASSWORD=password" on "docker run".

       You may also use "POSTGRES_HOST_AUTH_METHOD=trust" to allow all
       connections without a password. This is *not* recommended.

       See PostgreSQL documentation about "trust":
       https://www.postgresql.org/docs/current/auth-trust.html
```

The log currently shows that the user has no network access:

```
$ docker run clickhouse/clickhouse-server
/entrypoint.sh: neither CLICKHOUSE_USER nor CLICKHOUSE_PASSWORD is set, disabling network access for user 'default'
Processing configuration file '/etc/clickhouse-server/config.xml'.
Merging configuration file '/etc/clickhouse-server/config.d/docker_related_config.xml'.
Logging trace to /var/log/clickhouse-server/clickhouse-server.log
Logging errors to /var/log/clickhouse-server/clickhouse-server.err.log
```

Maybe we could improve it by sending this message to `stderr`.

Besides that, we became slightly closer to the standards.

Felixoid commented:
I'll update the docker docs in #74605 and on the https://hub.docker.com/r/clickhouse/clickhouse-server as well

Both `CLICKHOUSE_USER + CLICKHOUSE_PASSWORD` are already there, but it's worth to mention we have insecure fallback with `CLICKHOUSE_SKIP_USER_SETUP` now.

slvrtrn commented:
> There's another env parameter CLICKHOUSE_SKIP_USER_SETUP=1, the best fallback to the old behavior

Thanks, that helps!

Felixoid commented:
You are very welcome to review the docs' wording in https://github.com/ClickHouse/ClickHouse/pull/74605/files#diff-cb35cbd88e33b5c202774e3e47b97fb914a161f8a67e98d3c453ef57a56e81d1

Felixoid commented:
@darkleaf I see, you're a man of memes!

![image](https://github.com/user-attachments/assets/1a678a6a-7f21-45e0-af5f-ceb49458ea0f)

darkleaf commented:
@Felixoid 
![rn_image_picker_lib_temp_3f09c36b-610f-4cc3-ac02-98f4d7e8b842-tzmxpja3qfg97fzssexd6c8sic](https://github.com/user-attachments/assets/73883929-cf78-4673-864a-90b7cc06f26f)

darkleaf commented:
@Felixoid we have
```
   strategy:
      matrix:
        image:
          - redis:      "redis:7.4"
            clickhouse: "clickhouse/clickhouse-server:24.8"
          - redis:      "redis:7.4"
            clickhouse: "clickhouse/clickhouse-server:24.3"
          - redis:      "redis:6.2"
            clickhouse: "clickhouse/clickhouse-server:23.8"
        experimental: [false]
        include:
          - image:
              redis:      "redis:latest"
              clickhouse: "clickhouse/clickhouse-server:latest"
            experimental: true
```

Felixoid commented:
Touche

Slach commented:
@Felixoid why you choose to runtime change configuration files inside the container in `entrypoint.sh` Instead of just change `docker_related_config.xml`? These changes mixed declarative configuration approach on build time. 

These changes actually will just broke lot of installation, which will use read-only volumes for `/etc/clickhouse-server` or `/etc/clichouse-server/users.d/`

This is not make `clickhouse-server` docker installation more secure because most of the users just will decide `CLICKHOUSE_SKIP_USER_SETUP=1` and `default` without password user will still present

Hope, these changes not relates to the latest deepseek stupid installation mistakes ;(

alexey-milovidov commented:
@Felixoid, when the password for the default user is already set in one or another configuration file, it should not argue.

Felixoid commented:
> why you choose to runtime change configuration files inside the container in entrypoint.sh Instead of just change docker_related_config.xml? These changes mixed declarative configuration approach on build time.

Because there was a code that managed users, hardcoding defaults looked worse at the time. The purpose of the changes is pushing users to set up the user/password. 

> These changes actually will just break many installations, which will use read-only volumes for /etc/clickhouse-server or /etc/clichouse-server/users.d/

That was another point in favor of adding `CLICKHOUSE_SKIP_USER_SETUP` to make `default` as before. AFAIK, the RBAC user management is used for such cases. Unfortunately, not everyone is so advanced to set it up.

> This is not make clickhouse-server docker installation more secure because most of the users just will decide `CLICKHOUSE_SKIP_USER_SETUP=1` and `default` without password user will still present

Then it's on the user to leave the DB completely open, right?

> when the password for the default user is already set in one or another configuration file, it should not argue. Configuration files can be mapped to the container in numerous ways.

It's an interesting point that could be implemented. However, making it work is tricky because of the numerous ways.

The 100% working way would imply the patch to `clickhouse-extract-from-config` that would automatically read the users' configuration by an arg '--users'. Something like https://github.com/ClickHouse/ClickHouse/blob/0ae052180df6cee193a07ed629dda90cbfcab5de/programs/local/LocalServer.cpp#L475-L489

Then we can check the content of `users.default` for keys `password`, `password_sha256_hex`, `password_double_sha1_hex`, `ldap.server`, or `kerberos`.

azat commented:
>Then we can check the content of users.default for keys password, password_sha256_hex, password_double_sha1_hex, ldap.server, or kerberos.

I would say just check that the user is configured, regardless of the `password*`/`no_password`, since if he provides the config - it was the "explicit intention"

Felixoid commented:
It's a chicken-egg issue, kind of.

That means we need to compare the content of the `default` user to the "matrix" one. Copy `/etc/clickhouse-server/users.xml` into `/tmp` and compare the output to the one without `users.d`?

Felixoid commented:
It looks like the https://github.com/ClickHouse/ClickHouse/pull/75643 addresses all concerns, especially the case when `default` user is already changed somehow

azat commented:
I don't want to beat about the bush, but I kind of liked the way clickhouse images works before, no need to set any env variables - it just works (unlike mysql, where you need to go and read documentation for the image make it run)

Felixoid commented:
Everybody interesting in checking the fix from #75643 is working, you can check it as following:

```
cd ClickHouse # repo root
git fetch https://github.com/ClickHouse/ClickHouse.git
git checkout 183c840f10c9d869e5346357b3e7a6037247a9b4
docker buildx build --output=type=docker --platform=linux/amd64 --label=build-url=https://github.com/ClickHouse/ClickHouse/actions/runs/13187011934 --label=com.clickhouse.build.githash=183c840f10c9d869e5346357b3e7a6037247a9b4 --label=com.clickhouse.build.version= --build-arg=DIRECT_DOWNLOAD_URLS='https://s3.amazonaws.com/clickhouse-builds/PRs/75643/579c3023f7188f156e9a08320f522da467ccf2f9/package_release/clickhouse-client_25.2.1.1202_amd64.deb https://s3.amazonaws.com/clickhouse-builds/PRs/75643/579c3023f7188f156e9a08320f522da467ccf2f9/package_release/clickhouse-common-static_25.2.1.1202_amd64.deb https://s3.amazonaws.com/clickhouse-builds/PRs/75643/579c3023f7188f156e9a08320f522da467ccf2f9/package_release/clickhouse-server_25.2.1.1202_amd64.deb' --tag=clickhouse/clickhouse-server:head-amd64 --build-arg=VERSION='25.2.1.1' --progress=plain --file=docker/server/Dockerfile.ubuntu docker/server
```

You can check in advance if it works

pkit commented:
@Felixoid 
> Then it's on the user to leave the DB completely open, right?

Nobody leaves anything open. Users are mapped from config maps or volumes or files.
Obviously read-only for security reasons. And then all other users are managed from DDL.
That setup is completely destroyed by the PR here.

@Slach 
> Hope, these changes not relates to the latest deepseek stupid installation mistakes

They for sure are. The Wiz attention whoring was hilarious though.
The right course of action is to ignore. But you know...

Felixoid commented:
> That setup is completely destroyed by the PR here.

Does it mean you've checked the message above on how to check the follow-up PR?

What was the result?

Felixoid commented:
```
$ docker run --rm -v /etc/clickhouse-server/users.d/:/etc/clickhouse-server/users.d/:ro --name xxx1 clickhouse/clickhouse-server:head
Processing configuration file '/etc/clickhouse-server/config.xml'.
Merging configuration file '/etc/clickhouse-server/config.d/docker_related_config.xml'.
Logging trace to /var/log/clickhouse-server/clickhouse-server.log
Logging errors to /var/log/clickhouse-server/clickhouse-server.err.log
$ docker run --rm -v /tmp/wrong-dir:/etc/clickhouse-server/users.d/:ro --name xxx1 clickhouse/clickhouse-server:head
/entrypoint.sh: neither CLICKHOUSE_USER nor CLICKHOUSE_PASSWORD is set, disabling network access for user 'default'
/entrypoint.sh: line 156: /etc/clickhouse-server/users.d/default-user.xml: Read-only file system
$ docker run --rm --name xxx1 clickhouse/clickhouse-server:head
/entrypoint.sh: neither CLICKHOUSE_USER nor CLICKHOUSE_PASSWORD is set, disabling network access for user 'default'
Processing configuration file '/etc/clickhouse-server/config.xml'.
Merging configuration file '/etc/clickhouse-server/config.d/docker_related_config.xml'.
Logging trace to /var/log/clickhouse-server/clickhouse-server.log
Logging errors to /var/log/clickhouse-server/clickhouse-server.err.log
$ docker run --rm --name xxx1 -e CLICKHOUSE_SKIP_USER_SETUP=1 clickhouse/clickhouse-server:head
/entrypoint.sh: explicitly skip changing user 'default'
Processing configuration file '/etc/clickhouse-server/config.xml'.
Merging configuration file '/etc/clickhouse-server/config.d/docker_related_config.xml'.
Logging trace to /var/log/clickhouse-server/clickhouse-server.log
Logging errors to /var/log/clickhouse-server/clickhouse-server.err.log
$ docker run -it --rm -v /etc/clickhouse-server/users.d/:/etc/clickhouse-server/users.d/:ro --name xxx1 clickhouse/clickhouse-server:head bash
root@f973bb045d2a:/#
exit
```

maederm commented:
@Felixoid, could you add this change to the breaking changes section in the changelog?

maederm commented:
@Felixoid: I tested the [image](https://hub.docker.com/layers/clickhouse/clickhouse-server/25.1/images/sha256-351c0491ef4ff9c8705304336c9c38f77afa2b4dda2add0dad6a84b489d0b948), but in my use case it still breaks the CI pipeline running integration tests as it is started like `docker run --rm -p 127.0.0.1:9000:9000 clickhouse/clickhouse-server:latest`. Using `-e CLICKHOUSE_SKIP_USER_SETUP=1`  brought back the old behavior.

[Here ](https://www.jenkins.io/doc/book/pipeline/docker/#running-sidecar-containers) is an example of such a CI pipeline.

Felixoid commented:
Yes, the `-p 127.0.0.1:9000:9000` doesn't work as the `localhost` under the hood https://github.com/ClickHouse/ClickHouse/pull/74605/files#diff-cb35cbd88e33b5c202774e3e47b97fb914a161f8a67e98d3c453ef57a56e81d1R76

To address it, `--network=host` will help.

I'll take the documentation part from the #74605 and will update it on the docker hub ASAP

pkit commented:
> > That setup is completely destroyed by the PR here.
> 
> Does it mean you've seen the code from the message above on how to check the follow-up PR?
> 
> What was the result?

I'm not sure how to answer that one.
I have a perfectly secure (as it can get) flow.
Now I need to change it for no reason.
No matter the setup, I need to change it. Because it fails.
Obviously the easiest change for me was just to set `CLICKHOUSE_SKIP_USER_SETUP=1` to remove the strange attempt at "security" that just brings more headache. :)
In the future I think I will just return to building the image myself with a proper entrypoint.

Felixoid commented:
> No matter the setup, I need to change it. Because it fails.

With the `latest` image version, the user setup step is automatically omitted if the user `default` is changed in the `users.d`.

pkit commented:
> > No matter the setup, I need to change it. Because it fails.
> 
> With the `latest` image version, the user setup step is automatically omitted if the user `default` is changed in the `users.d`.

That's much better! Unfortunately I don't see which version of CH supports `--users` ...

Felixoid commented:
25.1 supports it, as does the latest master.

The rest of the active releases will receive updated entrypoint on Monday.

pkit commented:
Which one of the 25.1s? Lol
25.2 does not support it asa far as I'm concerned:
```
std::exception. Code: 1001, type: boost::wrapexcept<boost::program_options::unknown_option>, e.what() = unrecognised option '--users' (version 25.2.1.1840 (official build))
```

Felixoid commented:
the most recent one, 25.1.4.53, supports it

```
$ clickhouse-extract-from-config --version
std::exception. Code: 1001, type: boost::wrapexcept<boost::program_options::unknown_option>, e.what() = unrecognised option '--version' (version 25.1.4.53 (official build))
$ clickhouse-extract-from-config --version
std::exception. Code: 1001, type: boost::wrapexcept<boost::program_options::unknown_option>, e.what() = unrecognised option '--version' (version 25.1.4.53 (official build))
```

The master supports it too, that's how to check it:

```
$ bash -c "$(curl https://clickhouse.com)"
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  2911    0  2911    0     0  35461      0 --:--:-- --:--:-- --:--:-- 35938
ClickHouse binary clickhouse already exists. Overwrite? [y/N] y

Will download https://builds.clickhouse.com/master/amd64/clickhouse into clickhouse

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  121M  100  121M    0     0  12.4M      0  0:00:09  0:00:09 --:--:-- 12.5M

Successfully downloaded the ClickHouse binary, you can run it as:
    ./clickhouse

You can also install it:
sudo ./clickhouse install
$ ./clickhouse extract-from-config --help |& grep users
  --users                  Return values from users.xml config
$ ./clickhouse extract-from-config --version
std::exception. Code: 1001, type: boost::wrapexcept<boost::program_options::unknown_option>, e.what() = unrecognised option '--version' (version 25.2.1.2374 (official build))
```
